pkg/local_object_storage/writecache/flush.go

```
@@ -207,9 +207,6 @@ func (c *cache) flushFSTree(ignoreErrors bool) error {
 
 		err = c.flushObject(&obj, data)
 		if err != nil {
-			if ignoreErrors {
```

can you explain this change, please? what led to "the error should be returned as with Bolt" argument?

btw commit message has "Although there _ARE_ `ignoreErrors` flag"

---

I did this for this test: https://github.com/nspcc-dev/neofs-node/blob/8b65833d3651185fd9078e36b5acac73b1eaf762/pkg/local_object_storage/writecache/flush_test.go#L95-L96
It used to work because only the bolt was used and the error was not ignored there when flushing:
https://github.com/nspcc-dev/neofs-node/blob/8b65833d3651185fd9078e36b5acac73b1eaf762/pkg/local_object_storage/writecache/flush.go#L322-L324

---

Tricky one. The behavior of ignoring the error comes from b64b14eb54cac04a27049e9eb0ef6fc414a85d59 which is made quite some time after the 0b4c867ef1381a947b2821d9759e4985e12a5249. Semantically, I'd say that if we're to `ignoreErrors`, we should ignore all of them, flush is not guaranteed in this case (best-effort). Treating is a "ignore read errors only" is a possibility too, but maybe it'd be more problematic in real life. The flag itself is suspicious to me, but while we have it we probably better keep the behavior as it is now and then the test should be adjusted accordingly (try flushing with false/true to check for the difference?)

---

Here in [`SetMode`](https://github.com/nspcc-dev/neofs-node/blob/b3e19e2c00dedfd5d6620a62859567ac7885458c/pkg/local_object_storage/writecache/mode.go#L17), if I change `ignoreErrors` to `false`:
https://github.com/nspcc-dev/neofs-node/blob/b3e19e2c00dedfd5d6620a62859567ac7885458c/pkg/local_object_storage/writecache/mode.go#L21-L26
All tests will be passed without this change.

---

These flushers are very tricky in fact. The node should continue funtioning, but ignoring errors silently is bad. The most common flush error is "out of disk space" (the other is "your disk is broken forever"), continuing flushes in this case is not very productive.

---

> The flag itself is suspicious to me

as i remember, it was provided mostly for degraded cases when we think that disk is dying and we are trying to save (persist) as many objects as possible, so we ignore all problems about object X to be able to continue to object X+1

> Treating is a "ignore read errors only" is a possibility too

i guess this is the closest interpretation to what was planned to do

IMO, it is a good discussion for a separate issue, we also have a shard dump routine that also has some `ignoreError` things. can be fixed/discussed at a time

---

https://github.com/nspcc-dev/neofs-node/issues/3099